Feature/402-AsyncCallbackApiStt

adrien2p opened the pull request:
Add Asynchronous Callback API according to the documentation https://www.ibm.com/watson/developercloud/doc/speech-to-text/async.shtml#usage

- [x] npm test pass
- [x] register callback url
- [x] create recognitions job
- [x] get recognitions job status
- [x] get recognition job status
- [x] delete recognition job
...

Fixes #402

germanattanasio commented:
@adrien2p you don't need to use `[WIP]` in the commit message.

adrien2p commented:
@nfriedly could you review it ? i think it's ok i would like to now if you have some feedback.

nfriedly commented:
Looking now

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,34 @@ describe('speech_to_text_integration', function() {
+    });
+
+    it('deleteRecognitionJob()', function(done) {
+      speech_to_text.getRecognitionJob({ id: 'Job01' }, done);
```
This is calling the wrong method.

nfriedly replied:
Also, the job id should correspond to the one created earlier. (And, if necessary, the test should wait for it to finish before deleting it.)

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,34 @@ describe('speech_to_text_integration', function() {
+    });
+
+    it('getRecognitionJob()', function(done) {
+      speech_to_text.getRecognitionJob({ id: 'Job01' }, done);
```
Is this a valid job id?

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,34 @@ describe('speech_to_text_integration', function() {
     });
+
+    it('registerCallback()', function(done) {
+      const params = { callback_url: 'http://www.callback.fr/results', user_secret: 'ThisIsMySecret' };
```
(If it needs to have a specific response, we can add that with a PR to https://github.com/germanattanasio/watson-test-resources/)

adrien2p replied:
yes i could add it later

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+      const path = '/v1/register_callback?callback_url=http%3A%2F%2Fwww.callback.fr%2Fresults&user_secret=ThisIsMySecret';
+      const response = {
+        status: 200,
+        url: 'http://www.callback.fr/results'
```
I know this one is offline but please use an IBM-owned domain here as well - e.g. ibm.com or watson-test-resources.mybluemix.net

nfriedly replied:
Op, and you'll need to update the encoded one in the path also

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+      nock(service.url).persist().post(path).delay(100).reply(200, response);
+
+      const checkRes = function(err, res) {
+        assert.equal(JSON.stringify(err), JSON.stringify(null));
```
this should just be `assert.ifError(err); - https://nodejs.org/api/assert.html#assert_assert_iferror_value

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+
+      const checkRes = function(err, res) {
+        assert.equal(JSON.stringify(err), JSON.stringify(null));
+        assert.equal(JSON.stringify(res), JSON.stringify(response));
```
This should be `assert.deepEqual(res, response)` - https://nodejs.org/api/assert.html#assert_assert_deepequal_actual_expected_message

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+        url: 'http://www.callback.fr/results'
+      };
+
+      nock(service.url).persist().post(path).delay(100).reply(200, response);
```
You don't need to call `.persist()` here, that's for setting up a single response to handle multiple tests.

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+        assert.equal(JSON.stringify(res), JSON.stringify(response));
+      };
+
+      const params = { callback_url: 'http://www.callback.fr/results', user_secret: 'ThisIsMySecret' };
```
domain

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+    });
+
+    it('should create new recognitions job', function() {
+      const path = '/v1/recognitions?callback_url=http%3A%2F%2Fwww.callback.fr%2Fresults&event=recognitions.completed&user_token=ThisIsMySecret&results_ttl=60';
```
domain

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+        url: 'https://stream.watsonplatform.net/speech-to-text/api/v1/recognitions/4bd734c0-e575-21f3-de03-f932aa0468a0'
+      };
+
+      nock(service.url).persist().post(path).delay(100).reply(200, response);
```
persist

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+
+      const checkRes = function(err, res) {
+        assert.equal(JSON.stringify(err), JSON.stringify(null));
+        assert.equal(JSON.stringify(res), JSON.stringify(response));
```
asserts

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+      const params = {
+        audio: fs.createReadStream(__dirname + '/../resources/weather.wav'),
+        content_type: 'audio/l16;rate=41100',
+        callback_url: 'http://www.callback.fr/results',
```
domain

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+        ]
+      };
+
+      nock(service.url).persist().get(path).delay(200).reply(200, response);
```
persist

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+
+      const checkRes = function(err, res) {
+        assert.equal(JSON.stringify(err), JSON.stringify(null));
+        assert.equal(JSON.stringify(res), JSON.stringify(response));
```
asserts

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+        status: 'completed'
+      };
+
+      nock(service.url).persist().get(path).delay(200).reply(200, response);
```
persist

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+
+      const checkRes = function(err, res) {
+        assert.equal(JSON.stringify(err), JSON.stringify(null));
+        assert.equal(JSON.stringify(res), JSON.stringify(response));
```
asserts

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+  it('should delete a recognition job', function() {
+    const path = '/v1/recognitions/Job01';
+
+    nock(service.url).persist().delete(path).delay(200).reply(200);
```
persist

nfriedly commented on test/unit/test.speech_to_text.v1.js:
```diff
@@ -344,4 +344,147 @@ describe('speech_to_text', function() {
+    nock(service.url).persist().delete(path).delay(200).reply(200);
+
+    const checkRes = function(err, res) {
+      assert.equal(JSON.stringify(err), JSON.stringify(null));
```
assert

adrien2p commented:
Okay perfect i resolve that to night :) i hope it was good enougth :)

adrien2p commented:
@nfriedly i think it's good now

nfriedly commented:
So, when I checked out the code locally and ran the tests, the registerCallback() integration test fails, and then that failure cascades to kill most of the other integration tests. Looking at the docs, it seems that we need server-side support. I'm working on that now for the watson-test-resources server.

adrien2p commented:
I'm quite surprise cause i've crated a local server and provide localhost, with callback url which return the challenge_string, publicly with localtunel to really test callback register method ...

adrien2p commented:
i've only a free access on blumix, i'm really limited ...

nfriedly commented:
Ok, I added support to the test resources server, both for signed and unsigned requests. I updated the test code to hit the signed endpoint, and added a missing `done()` to the `createRecognitionJob()` test and now they all pass.

Looking at the logs from the server, it seems that the transcription is failing no matter what audio file I send, but I think that may be an unrelated service issue. Did that work for you?

adrien2p commented:
i'll can try on my local server exposed tomorrow, i go to bed now :) see you

nfriedly commented:
Figured it out: the audio has to be piped in rather than included as a param in the body. In other words, the audio *is* the body.

Of course, now the delete test is consistently failing...

adrien2p commented:
there is a problem on my blue mix account, i cant't have any credentials due to a problem, i've open a ticket. i try when i've access again

adrien2p commented:
ok, so i've pushed something but i can't try without credentials, so i wanting the ibm fix bluemix credentials serve

adrien2p commented:
The credentials service works again, so, i've tried to execute all processes and that work really well :)
```js
{ 
url: 'https://bwakwfcofs.localtunnel.me/stt-results',
status: 'created' 
}

{
reated: '2017-02-26T08:18:47.631Z',
id: '325795f0-fbfc-11e6-a8fd-5f28bc58ceb2',
url: 'https://stream.watsonplatform.net/speech-to-text/api/v1/recognitions/325795f0-fbfc-11e6-a8fd-5f28bc58ceb2',
status: 'processing' 
}

{ recognitions: 
   [ { created: '2017-02-26T08:18:47.631Z',
       id: '325795f0-fbfc-11e6-a8fd-5f28bc58ceb2',
       user_token: 'ThisIsMySecret',
       updated: '2017-02-26T08:18:56.974Z',
       status: 'completed' },
     { created: '2017-02-26T08:23:46.688Z',
       id: 'e4981000-fbfc-11e6-b73e-fdbb6821ee7d',
       user_token: 'ThisIsMySecret',
       updated: '2017-02-26T08:23:55.871Z',
       status: 'completed' } ] }

{ created: '2017-02-26T08:23:46.688Z',
  id: 'e4981000-fbfc-11e6-b73e-fdbb6821ee7d',
  updated: '2017-02-26T08:23:55.871Z',
  results: [ { result_index: 0, warnings: [Object], results: [Object] } ],
  status: 'completed' }
```

nfriedly commented on speech-to-text/v1.js:
```diff
@@ -91,6 +91,163 @@ SpeechToTextV1.prototype.version = 'v1';
+ * @param {string} params.callback_url - A URL to which callback notifications are to be sent
+ * @param {string} params.event - recognitions.started|recognitions.completed|recognitions.failed|recognitions.completed_with_results
+ * @param {string} params.user_token - The token allows the user to maintain an internal mapping between jobs and notification events
+ * @param {number} params.results_ttl - time to alive of the job result
```
event, token, and results are all optional, so they should be in [brackets]

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,42 @@ describe('speech_to_text_integration', function() {
+          audio: fs.createReadStream(__dirname + '/../resources/weather.wav'),
+          content_type: 'audio/l16;rate=41100',
+          callback_url: 'http://watson-test-resources.mybluemix.net/results',
+          user_token: 'ThisIsMySecret',
```
Please change this to something like `'my-arbitrary-identifier-1'`.

`uset_token` is an optional arbitrary string that you can set to easily tie the event callbacks to this original request. However, it should *not* be set to the same value as your secret, because that would compromise the secret.

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,42 @@ describe('speech_to_text_integration', function() {
       customization_id = null;
     });
+
+    describe('asynchronous api', function() {
```
It looks like you accidentally put this inside of the `describe('customization', function() {` section. Just move one of the `});`s from the end of the file to just above here, and then let `npm run autofix` clean up the indentation ;)

nfriedly commented on test/integration/test.speech_to_text.js:
```diff
@@ -380,5 +380,42 @@ describe('speech_to_text_integration', function() {
+      });
+
+      it('deleteRecognitionJob()', function(done) {
+        speech_to_text.deleteRecognitionJob({ id: jobId }, done);
```
Do you have a set of credentials, and does this test pass when you run it locally?

It looks for credentials in `test/resources/auth.js`, and if that file doesn't exist, then it skips all integration tests.

You can add them for just this service by creating the file with content like so:

```js
'use strict';

module.exports = {
  speech_to_text: {
    url: 'https://stream.watsonplatform.net/speech-to-text/api',
    username: '<your username>',
    password: '<your password>',
    headers: {
      'X-Watson-Learning-Opt-Out': 1,
      'X-Watson-Test': 1
    },
    version: 'v1'
  },
  // other services would go here
};
```

With that in place, you can run

    node ./node_modules/mocha/bin/_mocha /Users/nfriedly/node-sdk/test/integration/test.speech_to_text.js --grep "speech_to_text_integration asynchronous api "

(Note that this command expects you have already made the above change to move it outside of the customization describe.)

Note that with that file in place, the `npm test` command will try to run *all* integration tests, and the rest of them will fail due to missing credentials.  I'm going to fix that eventually, but in the meanwhile, you can just rename it to something else.

nfriedly commented on webpack.config.js:
```diff
@@ -4,6 +4,10 @@
+// With this configuration, the entire library can be bundled with Webpack. 
+// However, not all services support client-side usage.
+// See https://github.com/watson-developer-cloud/node-sdk/tree/master/examples/webpack for details
+
```
I think this was probably introduced by accident in a merge commit.

nfriedly commented:
Oh, and I just fixed the webpack config in master, so the tests should start passing again if you merge that in

adrien2p commented:
Hi, i'll work on it to night, yes for me the delete method works, i've tried to call the api to delete a created job and that's works fine. So i think it's because manually the time between creation, check and delete is long enough, so i think in the test i must wait before trying to delete.

nfriedly commented:
Thanks @adrien2p! 

This should be released as v2.22.0 by CI shortly.

adrien2p commented:
My pleasure, i waiting for the next contribution :)
